### PR TITLE
Adds a .ResolvedTarget Property to File-System Items to Reflect a Symlink's Target as FileSystemInfo

### PR DESCRIPTION
**Pull request by hammy275**
# PR Summary

This PR addresses issue #13366 , adding the ability for users to access a `ResolvedTarget` property on file and folder items, which returns a string representation of a path, depending on the type of file/folder:

* If the file/folder is not a symbolic link, `ResolvedTarget` will return a string referring to itself.
* If the file/folder is a symbolic link that points to a file/folder that currently exists, `ResolvedTarget` will return a string referring to the file/folder that the symbolic link points to.
* If the file/folder is a symbolic link that points to a file that does not exist, `ResolvedTarget` returns a string referring to the broken destination.

## PR Context

This PR mainly addresses the problems described in #13366 , mainly that there is no easy way to resolve the path of a file if said file has a chance of being a symbolic link (and as a result, points to a different file in the end). This PR addresses this with the `ResolvedTarget` property, which points to "the actual file" whether the file being referenced is a symbolic link or not.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/8505
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by kmccl922** [comment]
Stop sending me these damn messages I'm going to report you'll ass 

Sent from Yahoo Mail on Android 
 
  On Thu, Nov 25, 2021 at 9:02 PM, ***@***.***> wrote:   


This pull request has been automatically marked as Review Needed because it has been there has not been any activity for 7 days.
Maintainer, please provide feedback and/or mark it as Waiting on Author

**Comment by iSazonov** [comment]
@hammy3502 Please look test fails.

@mklement0 Could you please review? Is this addressed a behavior you requested?

**Comment by hammy3502** [comment]
My apologies for the unit tests! The PR should be good to go now. The one failing unit test here looks to be from upstream from when I last rebased, rather than this PR.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (anamnavi) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8154]:
```suggestion
```

**Review by anamnavi — APPROVED** [review]
LGTM, just had one recommendation to remove extra newline

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8147]:
```suggestion
                } else if (fileSysInfo.LinkTarget is null) 
```

**Review by iSazonov — APPROVED** [review]

**Comment by hammy3502** [comment]
It looks like the failed tests are from outside the scope of this PR, so we should still be good to merge

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8139]: The issue is asking for a full path (return string result), but you made it return `FileInfo` or `DirectoryInfo` instead, why not returning string as requested in the issue?
The existing properties `Target` and `LinkTarget` both return a string, so I think it's better to do the same for `ResolvedTarget` property.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8139]: What is a negative effects from the throw? Should we really mask the exceptions?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8143]: What is a negative effects from the throw? Should we really mask the exceptions?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8139]: While do we need to limit the method? PowerShell does FileInfo.ToString() for us.

Comment (daxian-dbw) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8139]: We should be consistent with `Target` and `LinkTarget`, because the name `ResolvedTarget` indicates the same parameter type.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8143]: When calling the method on a file on a USB drive or micro sd drive, it will fail because the underlying filesystem doesn't support reparsing point. However, the behavior in that case should be returning the full path of the file.

Actually, this scenario should be handled, meaning when exception happens, it should decide when to return the `FullName` of the current `FileSystemInfo`, and when to return `null`.

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8143]: This would weird behavior in formatting output - if user see Target and LinkType  properties are "empty" but ResolvedTarget is FullName.
And if I remember correctly .Net ResolveLinkTarget can return partially resolved path. https://github.com/dotnet/runtime/blob/12a8819eee9865eb38bca6c05fdece1053102854/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs#L621-L627

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8139]: > We should be consistent

Why? .Net object is more useful than string. And PowerShell will convert to string itself which it does in many scenarios.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8143]: The behavior I describe is what was asked from the issue:

> If the item at hand isn't a symlink / reparse point, simply report its own full path - this allows unified treatment of symlinks and non-symlinks, as a simple way to get an item's full path.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8139]: Well, if there is disagreement on the parameter type, then let's have it reviewed by the committee.

**Comment by daxian-dbw** [comment]
@PowerShell/powershell-committee We need the committee to review the issue and PR on 2 behaviors of the proposed `ResolvedTarget` property:
1. when the file or directory is not a symlink / reparse point, whether it should repot its own full path, or return `null`. See https://github.com/PowerShell/PowerShell/pull/16490#discussion_r769134270 for the discussion.
2. whether the parameter type of the proposed property to be `FileSystemInfo` or `string`. see https://github.com/PowerShell/PowerShell/pull/16490#discussion_r769138284 for the discussion.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8139]: Since you mentioned PSPath... :-) Last days I am thinking about FileSystem Provider V2 and related issues although it is more correct to say the problems are more related to _path conversions_. And now I'm more and more inclined to think that PSPath should be an object and not just a string, and the other ETS properties could be based on that, and it would great to inject this in PSObject itself. 
This is much more in the spirit of PowerShell, opens the door to fixing many long-standing bugs (by experimental feature(s)), improving engine and provider performance, and making cross-provider operations easier to implement.

**Comment by hammy3502** [comment]
Apologies for the slow response! I'll do my best to have the changes back within a couple days!

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (hammy3502) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8147]: After removing the else block (the block on line 8147), my tests fail when `.ResolvedTarget` attempts to resolve for a target that isn't a symlink. Based on the issue, it says non-symlinks should resolve to themselves, however without that block, it resolves to `null`.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (hammy3502) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8143]: I'm about to push a commit that wraps a try-catch around the `ResolveLinkTarget` call and returning `null`, however I'm not sure in what situation `ResolveLinkTarget` would throw an exception and want to return `FullName`. It cleanly returns `null` when encountering something that isn't a symlink, and flash drives (by flashdrives, I'm assuming you mean the usage of FAT32/exFAT here) don't seem to support symlinks in any capacity. Am I missing something here? Thanks!

Comment (daxian-dbw) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I think we want to return `FullName` only if the `FileSystemInfo` is just a regular file/directory instead of a link.

Looking more into it, I found on portable drive with exFAT file system, `ResolveLinkTarget(false)` throws while `ResolveLinkTarget(true)` doesn't throw ([see code here](https://github.com/dotnet/runtime/blob/09c1a1f7b0c477890b04912d8dd4f742f80faffc/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs#L417-L425)). So, maybe we don't really need the `try / catch (IOException)` here.

So, maybe we can do the following:
1. if `ResolveLinkTarget(true)` returns `non-null` value, just return that value. (I see you check on `symbTarget.Exists` here, but could it even be possible that the method returns a non-null value, but it doesn't exist?)
2. if `ResolveLinkTarget(true)` returns `null` value, assuming that means it's not a link, and thus return the `FullName` or the current `FileSystemInfo` depending on what we want for return type.
3. if `ResolveLinkTarget(true)` throws, then we assume that means it could be link, but something is wrong when resolving it. We just let the exception propagate in this case.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Yeah, that was the intent of the ask. Thanks

**Comment by daxian-dbw** [comment]
@hammy3502 Thanks for your reply. I responded to your comments, and let's wait on the committee to decide the return type.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (hammy3502) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8143]: Based on the docstring for `ResolveLinkTarget`, it looks like we do need to check if it exists:

> Returns: A System.IO.FileSystemInfo instance if the link exists, independently if the target exists or not

Besides that, the commit I mentioned earlier and the ones I just pushed should match the requested behavior!

**Comment by hammy3502** [comment]
@daxian-dbw Thank you so much! I handled the two threads you responded to, and I'll be eagerly awaiting a response from the committee! :)

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this:

1. We agreed that `.ResolvedTarget` should always have a value that is the full path regardless if the item is a link.  Users should rely on `.Attributes` to determine if an item is a link or not.  By having this property always have a value, users won't need to check if it's `$null` and can also make it easier to format without conditionals.

2. We also agreed that for consistency and also it's easy to convert the string path to a FileSystemInfo object and most users may not want to incur the additional memory to allocate that object, it makes sense for `.ResolvedTarget` to be a string.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8150]:
```suggestion
                try
                {
                    FileSystemInfo symbTarget = fileSysInfo.ResolveLinkTarget(true);
                    return symbTarget?.FullName;
                }
                catch (IOException)
                {
                    return null;
                }
```

**Review by iSazonov — DISMISSED** [review]

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8139]:
```suggestion
        public static string ResolvedTarget(PSObject instance)
```

Comment (hammy3502) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I pushed a change that should resolve this. However, with the code suggested earlier, we return the path to the destination of a broken symbolic link. Should this remain the case? Thanks!

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I agree to follow .Net behavior.

File: brokenLinkedDir
Comment (iSazonov) [inline brokenLinkedDir:1]: Please remove these temp files.

File: test/powershell/Modules/Microsoft.PowerShell.Management/FileSystemProviderExtended.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/FileSystemProviderExtended.Tests.ps1:686]: Please use `Should -BeTrue`.
Below too.

**Review by iSazonov — COMMENTED** [review]
@hammy3502 Please fix tests.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8151]: @hammy3502 As we discussed in https://github.com/PowerShell/PowerShell/pull/16490#discussion_r777764971, I think we don't need the try/catch block anymore given that `ResolveLinkTarget(true)` does not throw exception on a portable drive with `exFAT` file system (`ResolveLinkTarget(false)` throws but `ResolveLinkTarget(true)` doesn't).

Quoted it here again:
> So, maybe we can do the following:
>1. if `ResolveLinkTarget(true)` returns `non-null` value, just return that value. (I see you check on `symbTarget.Exists` here, but could it even be possible that the method returns a non-null value, but it doesn't exist?)
>2. if `ResolveLinkTarget(true)` returns `null` value, assuming that means it's not a link, and thus return the `FullName` or the current `FileSystemInfo` depending on what we want for return type.
>3. if `ResolveLinkTarget(true)` throws, then we assume that means it could be link, but something is wrong when resolving it. We just let the exception propagate in this case.

Even if `linkTarget` points to a non-existing item (somehow the target gets deleted), we should just return the full path to the target. Since .NET doesn't check on the existence of the target item before returning, we can follow the same behavior. So, the code within the try block looks good to me. I think you just need to remove the try/catch block.

**Review by daxian-dbw — COMMENTED** [review]
@hammy3502 Thanks for updating the changes! We are very closed to be ready :)

Please fix the tests. Let me know if you need help on fixing them.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: @hammy3502 about your question, please take a look at https://github.com/PowerShell/PowerShell/pull/16490#discussion_r782800772

File: brokenLinkedDir
Comment (daxian-dbw) [reply brokenLinkedDir:1]: Yeah, those files look like for manual testing only. Please remove them from the PR.

**Comment by hammy3502** [comment]
All of the changes should be resolved now! :)

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@hammy3502 Please rebase to fix CI-Static.

**Comment by iSazonov** [comment]
@hammy3502 Please update the PR description to reflect new logic is implemented and open new issue in Documentation repository https://github.com/MicrosoftDocs/PowerShell-Docs.

**Comment by hammy3502** [comment]
@iSazonov I've updated the PR description, however I can't seem to find a page to propose any needed documentation changes (there doesn't appear to be any pages for `FileInfo`/`DirectoryInfo` that I could find). Are there any pages you'd recommend be updated so I can open an issue? Thank you so much!

**Comment by iSazonov** [comment]
> I've updated the PR description

I don't see all cases covered in https://github.com/PowerShell/PowerShell/pull/16490#discussion_r782800772

> I can't seem to find a page to propose any needed documentation changes

Since it is new addition I think it should be documented. If PowerShell documentation hasn't appropriate pages you can new doc issue as "idea" in doc repo.

**Comment by iSazonov** [comment]
@daxian-dbw How could we re-run `Update cgmanifest CI`?

**Comment by daxian-dbw** [comment]
I have no idea. I guess it's stuck for some reason, and we may have to close the PR and then re-open.

**Comment by iSazonov** [comment]
@daxian-dbw Thanks!

This cgmanifest is reminiscent of the Files.wxs story. At first we had to update it manually every time. After that exhausted everyone, the update became automatic. 😄 I hope to see auto update soon!

**Comment by iSazonov** [comment]
@daxian-dbw The action was finished but not reported again. 😟

**Comment by iSazonov** [comment]
Adding the ability for users to access a ResolvedTarget property on file and folder items, which returns a string representation of a path, depending on the type of file/folder:

- If the file/folder is not a symbolic link, ResolvedTarget will return a string referring to itself.
- If the file/folder is a symbolic link that points to a file/folder that currently exists, ResolvedTarget will return a string referring to the file/folder that the symbolic link points to.
- If the file/folder is a symbolic link that points to a file that does not exist, ResolvedTarget returns a string referring to the broken destination.

**Comment by iSazonov** [comment]
@hammy3502 Thanks for your contribution! Please open doc issue.

**Comment by iSazonov** [comment]
@hammy3502 Thanks for your contribution!

**Pull request merged**
